atc/policy/opa/opa.go

```diff
@@ -33,59 +33,67 @@ type opaInput struct {
 	Input policy.PolicyCheckInput `json:"input"`
 }
 
+type opaOuptut struct {
+	Allowed *bool    `json:"allowed,omitempty"`
+	Reasons []string `json:"reasons,omitempty"`
+}
+
 type opaResult struct {
-	Result *bool `json:"result,omitempty"`
+	Result *opaOuptut `json:"result,omitempty"`
 }
 
 type opa struct {
 	config OpaConfig
 	logger lager.Logger
 }
 
-func (c opa) Check(input policy.PolicyCheckInput) (bool, error) {
+func (c opa) Check(input policy.PolicyCheckInput) (policy.PolicyCheckOutput, error) {
 	data := opaInput{input}
 	jsonBytes, err := json.Marshal(data)
 	if err != nil {
-		return false, err
+		return policy.FailedPolicyCheck(), nil
 	}
 
 	c.logger.Debug("opa-check", lager.Data{"input": string(jsonBytes)})
 
 	req, err := http.NewRequest("POST", c.config.URL, bytes.NewBuffer(jsonBytes))
 	if err != nil {
-		return false, err
+		return policy.FailedPolicyCheck(), nil
 	}
 	req.Header.Set("Content-Type", "application/json")
 
 	client := &http.Client{}
 	client.Timeout = c.config.Timeout
 	resp, err := client.Do(req)
 	if err != nil {
-		return false, err
+		return policy.FailedPolicyCheck(), err
 	}
 	defer resp.Body.Close()
 
 	statusCode := resp.StatusCode
 	if statusCode != http.StatusOK {
-		return false, fmt.Errorf("opa returned status: %d", statusCode)
+		return policy.FailedPolicyCheck(), fmt.Errorf("opa returned status: %d", statusCode)
 	}
 
 	body, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return false, fmt.Errorf("opa returned no response: %s", err.Error())
+		return policy.FailedPolicyCheck(), fmt.Errorf("opa returned no response: %s", err.Error())
 	}
 
 	result := &opaResult{}
 	err = json.Unmarshal(body, &result)
 	if err != nil {
-		return false, fmt.Errorf("opa returned bad response: %s", err.Error())
+		return policy.FailedPolicyCheck(), fmt.Errorf("opa returned bad response: %s", err.Error())
 	}
 
 	// If no result returned, meaning that the requested policy decision is
 	// undefined OPA, then consider as pass.
 	if result.Result == nil {
-		return true, nil
+		return policy.PassedPolicyCheck(), nil
 	}
```

Comment:
If someone writes a policy that only returns `reasons` users get an EOF error via fly or no message at all on the build page. We find the following in the Concourse logs
```
2020/08/20 14:56:03 http: panic serving 172.24.0.1:58766: runtime error: invalid memory address or nil pointer dereference
goroutine 196255 [running]:
net/http.(*conn).serve.func1(0xc0007b52c0)
	/usr/local/go/src/net/http/server.go:1800 +0x149
panic(0x34c6ce0, 0x5105c80)
	/usr/local/go/src/runtime/panic.go:975 +0x41b
github.com/concourse/concourse/atc/policy/opa.opa.Check(0xc000056012, 0x2a, 0x12a05f200, 0x3e85d20, 0xc0007987e0, 0x37e7ad1, 0x9, 0xc0000520b7, 0x3, 0x37e6c70, ...)
	/src/atc/policy/opa/opa.go:96 +0xee8
github.com/concourse/concourse/atc/policy.(*Checker).Check(0xc000798900, 0x37e7ad1, 0x9, 0xc0000520b7, 0x3, 0x37e6c70, 0x9, 0xc0005d4b80, 0x3, 0x37eb46b, ...)
	/src/atc/policy/checker.go:166 +0x11b
github.com/concourse/concourse/atc/api/policychecker.(*checker).Check(0xc0007f6038, 0x37eb46b, 0xb, 0x3e85ea0, 0xc000bc8230, 0xc000945500, 0xc000824600, 0x0, 0x0, 0x0, ...)
	/src/atc/api/policychecker/checker.go:77 +0xc6b
github.com/concourse/concourse/atc/api/policychecker.policyCheckingHandler.ServeHTTP(0x3e85d20, 0xc0003fcf60, 0x3e11300, 0xc00085c210, 0x37eb46b, 0xb, 0x3e0c020, 0xc0007f6038, 0x3e67fa0, 0xc000bc81c0, ...)
	/src/atc/api/policychecker/handler.go:38 +0xf4
github.com/concourse/concourse/atc/api/auth.checkAuthHandler.ServeHTTP(0x3e10f00, 0xc0003bc8c0, 0x3e2f0a0, 0x527beb0, 0x3e67fa0, 0xc000bc81c0, 0xc000945500)
	/src/atc/api/auth/check_authentication_handler.go:28 +0x92
github.com/concourse/concourse/atc/wrappa.VersionedHandler.ServeHTTP(0x37e6c70, 0x9, 0x3e10dc0, 0xc0003a7700, 0x3e67fa0, 0xc000bc81c0, 0xc000945500)
	/src/atc/wrappa/versioned_handler.go:14 +0xa7
github.com/concourse/concourse/atc/api/accessor.(*accessorHandler).ServeHTTP(0xc0008115c0, 0x3e67fa0, 0xc000bc81c0, 0xc000945400)
	/src/atc/api/accessor/handler.go:64 +0x478
github.com/NYTimes/gziphandler.GzipHandlerWithOpts.func1.1(0x3e67b20, 0xc0006670a0, 0xc000945400)
	/go/pkg/mod/github.com/!n!y!times/gziphandler@v1.1.1/gzip.go:336 +0x2ba
net/http.HandlerFunc.ServeHTTP(0xc000871290, 0x3e67b20, 0xc0006670a0, 0xc000945400)
	/usr/local/go/src/net/http/server.go:2041 +0x44
github.com/bmizerany/pat.(*PatternServeMux).ServeHTTP(0xc0003c0c60, 0x3e67b20, 0xc0006670a0, 0xc000945400)
	/go/pkg/mod/github.com/bmizerany/pat@v0.0.0-20170815010413-6226ea591a40/mux.go:117 +0x313
github.com/concourse/concourse/atc/api/auth.csrfValidationHandler.ServeHTTP(0x3e0bbc0, 0xc0003c0c60, 0x3e8c800, 0xc000358ba8, 0x3e67b20, 0xc0006670a0, 0xc000945400)
	/src/atc/api/auth/csrf_validation_handler.go:60 +0x60b
net/http.(*ServeMux).ServeHTTP(0xc00087d6c0, 0x3e67b20, 0xc0006670a0, 0xc000945400)
	/usr/local/go/src/net/http/server.go:2416 +0x1a6
github.com/concourse/concourse/atc/api/auth.WebAuthHandler.ServeHTTP(0x3e0fa40, 0xc00087d6c0, 0x3e8c800, 0xc000358ba8, 0x3e67b20, 0xc0006670a0, 0xc000945400)
	/src/atc/api/auth/web_auth_handler.go:55 +0x300
github.com/concourse/concourse/atc/wrappa.SecurityHandler.ServeHTTP(0x35031d2, 0x4, 0x3e10d80, 0xc0007b3a60, 0x3e67b20, 0xc0006670a0, 0xc000945400)
	/src/atc/wrappa/security_handler.go:17 +0x1af
github.com/concourse/concourse/atc/wrappa.LoggerHandler.ServeHTTP(0x3e85d20, 0xc0003fcf60, 0x3e11520, 0xc0007b3a80, 0x3e67b20, 0xc0006670a0, 0xc0007e9700)
	/src/atc/wrappa/logger_handler.go:20 +0x2fb
net/http.serverHandler.ServeHTTP(0xc0001ca380, 0x3e67b20, 0xc0006670a0, 0xc0007e9700)
	/usr/local/go/src/net/http/server.go:2836 +0x20f
net/http.(*conn).serve(0xc0007b52c0, 0x3e6fda0, 0xc000403080)
	/usr/local/go/src/net/http/server.go:1924 +0x173a
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2962 +0x931
```

This is definitely an edge-case, curious what your thoughts are? I was thinking we could add a case around here to check that `result.Result.Allowed == nil` as well? I think that was the original intention of this `if`-case.

Reply:
Looking at the original code:
```go
	type opaResult struct {
		Result *bool `json:"result,omitempty"`
	}
...
	if result.Result == nil {
		return policy.PassedPolicyCheck(), nil
	}

```
The intention was definitely to check if a bool was passed back. I think we should still do this check. The `if` in it's current state does not do this. I think it will always pass if `json.Unmarshal` was successful.

Reply:
Yeah that makes sense. I thought that a `bool` would deserialize to a default of `false` as it does in some other languages but if it ends up as `nil` then we should probably check for that.

I'm unsure of what the standard is for OPA integrations about how strict the "API" of what is returned should be. Personally, I think that `result.Result` being `nil` as well as the response not containing an `allowed` field should also be treated as a failed check. If a check has been sent to a policy checker, that means an operator has specifically configured Concourse to do so for that action & http method so the policy checker (OPA) should *specifically* say that something is allowed rather than Concourse assuming things are allowed by default.

It's easy enough to put `default decision = { "allowed": "true" }` in your Rego and it prevents the potential foot-cannon of you configuring OPA wrongly and it just returns `{}` for everything which Concourse then inteprets as allowed due to that being the default. In this situation I think returning something like:
```
return policy.FailedPolicyCheck(), fmt.Errorf("opa returned invalid response: %+v", result.Result)
```
may be a more defensive approach. I can easily add some additional tests to cover this if we think that this approach is worthwhile.

Reply:
> I thought that a bool would deserialize to a default of false 

It does in Go, but not in this case because the bool is actually a pointer to a bool. If it was just a `bool` instead of `*bool` then you'd get the expected result of `result.Result == false`. I'm guessing this was done in case OPA returned `{}`. 

I think it's worth seeing if we can turn this pointer into a bool instead. I think it makes sense to have it as a pointer if we were going to `json.Marshal(opaOutput)` but I don't think we ever will. @evanchaoli is there a case/reason for keeping `opaOutput.Allowed` as a `*bool` instead of a `bool`?

> Personally, I think that result.Result being nil as well as the response not containing an allowed field should also be treated as a failed check

I agree with this 👍 I'd be happy to go with that approach, I think it's safer. I checked the initial OPA PR and didn't see this mentioned anywhere.

Reply:
Actually if we made it a `bool` rather than `*bool` we'll not be able to differentiate between "didn't supply allowed and defaulted to false"  and "allowed was false" so might make it harder to alert an operator about what was wrong with their rules.

Thinking about the user-facing side of this - the error here is provided back to the user verbatim. Would it be preferable return an error just saying `opa returned invalid response` and log the actual response we got at INFO level or is it okay to display errors messages like `opa returned invalid response: &{Allowed:<nil> Reasons:[]}` to users?

Reply:
I think the latter would be fine, showing the full error to the user. It would not be helpful for users but it will be helpful for operators. Especially if the user reports the error to the operator, the operator doesn't have the take the extra step of digging through logs.